runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/PubsubIOTranslator.java

```diff
@@ -32,32 +34,25 @@
   /**
    * Implements PubsubIO Read translation for the Dataflow backend.
    */
-  public static class ReadTranslator<T> implements TransformTranslator<PubsubIO.Read.Bound<T>> {
+  public static class ReadTranslator<T> implements TransformTranslator<PubsubUnboundedSource<T>> {
     @Override
     @SuppressWarnings({"rawtypes", "unchecked"})
     public void translate(
-        PubsubIO.Read.Bound transform,
+        PubsubUnboundedSource<T> transform,
         TranslationContext context) {
       translateReadHelper(transform, context);
     }
 
     private <T> void translateReadHelper(
-        PubsubIO.Read.Bound<T> transform,
+        PubsubUnboundedSource<T> transform,
         TranslationContext context) {
-      if (!context.getPipelineOptions().isStreaming()) {
-        throw new IllegalArgumentException(
-            "PubsubIO.Read can only be used with the Dataflow streaming runner.");
-      }
+      Preconditions.checkState(context.getPipelineOptions().isStreaming(),
+          "PubsubIOTranslator.ReadTranslator is only for streaming pipelines.");
 
       context.addStep(transform, "ParallelRead");
       context.addInput(PropertyNames.FORMAT, "pubsub");
-      if (transform.getTopic() != null) {
-        context.addInput(PropertyNames.PUBSUB_TOPIC, transform.getTopic().asV1Beta1Path());
```

Comment:
I really don't like it.
The batch direct runner  creates the subscription on one worker when the pipeline begins data processing, cleans up when done.
The streaming Google Cloud Dataflow runner creates the subscription on dax when the job is setup, cleans up when the job is terminated.
The streaming Java-only runner would create it when the pipeline graph is constructed, and never cleans it up.
The lack of cleanup plus phase difference of creation time all feels wrong...

... but I'll do it anyway.